Xiaomi MiIO Fan: Xiaomi Air Humidifier integration

syssi opened the pull request:
## Description:

Device support for the Xiaomi Air Humidifier. 

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io/pull/4829

## Example entry for `configuration.yaml` (if applicable):
```yaml
fan:
  - platform: xiaomi_miio
    name: Xiaomi Air Humidifier
    host: 192.168.130.72
    token: 2b00042f7481c7b056c4b410d28f33cf
    model: zhimi.humidifier.v1
```

syssi committed:
Device support for the Xiaomi Air Humidifier.

syssi committed:
Requirements updated.

syssi committed:
"continuation line under-indented for visual indent" fixed.

syssi committed:
Make hound happy.

syssi committed:
Inadvertently added light.xiaomi_miio component removed from PR.

syssi committed:
Service descriptions added.

syssi committed:
One of the pylint errors fixed.

syssi commented:
I'm unsure about

```
R:396, 4: Method could be a function (no-self-use)
R:401, 4: Method could be a function (no-self-use)
R:406, 4: Method could be a function (no-self-use)
R:411, 4: Method could be a function (no-self-use)
R:416, 4: Method could be a function (no-self-use)
```

It's safe to remove "self"?

syssi committed:
Redundancy removed.

syssi committed:
pylint: disable=no-self-use added. The method signature is important here.

syssi committed:
Pylint fixed.

syssi committed:
Use a unique data key per domain.

rytilahti commented on homeassistant/components/fan/services.yaml:
```diff
@@ -120,8 +120,76 @@ xiaomi_miio_set_led_brightness:
+      description: Name of the xiaomi miio entity.
+      example: 'fan.xiaomi_miio_device'
+    humidity:
+      description: 	Target humidity. Allowed values are 30, 40, 50, 60, 70 and 80.
```
Unnecessary whitespace.

rytilahti commented on homeassistant/components/fan/services.yaml:
```diff
@@ -120,8 +120,76 @@ xiaomi_miio_set_led_brightness:
+      example: 'fan.xiaomi_miio_device'
+
+xiaomi_miio_set_extra_features:
+  description: Set the extra features.
```
This could be more elaborate, what are the extra features and what are the known values below?

rytilahti commented on homeassistant/components/fan/services.yaml:
```diff
@@ -120,8 +120,76 @@ xiaomi_miio_set_led_brightness:
       example: 1
+
+xiaomi_miio_set_auto_detect_on:
+  description: Turn the auto detect on.
```
Auto detect and brightness are relevant?

syssi replied:
Fixed.

rytilahti commented on homeassistant/components/fan/xiaomi_miio.py:
```diff
@@ -74,45 +152,95 @@
+
+SERVICE_SCHEMA_EXTRA_FEATURES = AIRPURIFIER_SERVICE_SCHEMA.extend({
+    vol.Required(ATTR_FEATURES):
+        vol.All(vol.Coerce(int), vol.Range(min=0))
```
Does range work when only the min value is given?

rytilahti commented on homeassistant/components/fan/xiaomi_miio.py:
```diff
@@ -74,45 +152,95 @@
+    else:
+        _LOGGER.error(
+            'Unsupported device found! Please create an issue at '
+            'https://github.com/rytilahti/python.miio/issues '
```
Typo in the URL, should this be reported to homeassistant though (considering that the backend library may already support the device)? Just a thought though, this is fine for me either way.

syssi replied:
It would be a feature request. That's why I don't want to mention homeassistant here. I added a link to my development repository now. So we can keep the python-miio repo clean for protocol issues.

rytilahti commented on homeassistant/components/fan/xiaomi_miio.py:
```diff
@@ -216,28 +329,152 @@ def _try_command(self, mask_error, func, *args, **kwargs):
+            ATTR_TURBO_MODE_SUPPORTED: None,
+            ATTR_SLEEP_MODE: None,
+            ATTR_AUTO_DETECT: None,
+        })
```
Would it make sense to have a list of these attrs and simply do `_state_attrs = {attr: None for attr in available_attributes}` and set the model attribute afterwards?

rytilahti commented on homeassistant/components/fan/xiaomi_miio.py:
```diff
@@ -250,30 +487,53 @@ def async_update(self):
+                ATTR_EXTRA_FEATURES: state.extra_features,
+                ATTR_TURBO_MODE_SUPPORTED: state.turbo_mode_supported,
+                ATTR_AUTO_DETECT: state.auto_detect,
+            })
```
The same idea that I elaborated above could be use here too, maybe the `available_attributes` could be a map of (attribute_name, attribute_getter)` or similar?

syssi replied:
Cool! I will care about.

rytilahti commented on homeassistant/components/fan/xiaomi_miio.py:
```diff
@@ -297,67 +560,236 @@ def speed(self):
+            return
+
+        yield from self._try_command(
+            "Turning the learn mode of the miio device off failed.",
```
Auto detect versus learn mode.

rytilahti reviewed (commented):
From a quick review it looks good! I left a couple of comments how it could be improved.

syssi committed:
Review incorporated.

syssi committed:
Map of available attributes added.

syssi committed:
Pylint fixed.
Attribute "volume" added.

rytilahti commented on homeassistant/components/fan/xiaomi_miio.py:
```diff
@@ -84,6 +84,57 @@
+    ATTR_BUZZER: 'buzzer',
+    ATTR_LED_BRIGHTNESS: 'led_brightness',
+    ATTR_SLEEP_MODE: 'sleep_mode',
+})
```
Considering minimum python version being 3.5 now, I think the following would also work to simplify it a bit more:
```
AVAILABLE_ATTRIBUTES_AIRPURIFIER = {**AVAILABLE_ATTRIBUTES_AIRPURIFIER_COMMON,
                                    ATTR_BUZZER: 'buzzer',
}
```

syssi committed:
Don't use the support flag bit mask as model identifier.
Determine support features and attributes at the constructor.
Use starred expressions at dicts instead of copies.

syssi committed:
Blank line removed.

syssi committed:
Use Async / await syntax.

syssi committed:
Make hound happy.

syssi commented on homeassistant/components/fan/xiaomi_miio.py:
```diff
@@ -196,50 +361,140 @@ def device_state_attributes(self):
-    @asyncio.coroutine
-    def async_turn_on(self: ToggleEntity, speed: str = None, **kwargs) -> None:
-        """Turn the fan on."""
+    async def async_turn_on(self: ToggleEntity, speed: str = None,
```
Should I remove the return type (ToggleEntity) here? `self._device` isn't part of the ToggleEntity. pycharm is complaining about.

rytilahti replied:
Yes, there is no need to add typing hints for `self`.

syssi committed:
Xiaomi Air Humidifier CA support added.

syssi committed:
Duplicate method removed.

syssi committed:
Air Purifier V3 support added.

syssi commented:
I'm abusing the system property supported_features at the moment. This will be fixed soon.

syssi committed:
Don't abuse the system property supported_features anymore.

syssi committed:
python-miio version bumped.

syssi committed:
Clean-up.

syssi committed:
Additional supported features refactoring completed.

syssi committed:
Merge branch 'dev' into feature/xiaomi-miio-airhumidifier

syssi committed:
Additional supported features renamed properly.

syssi committed:
Unique id added.

syssi committed:
Device unavailable handling improved.

syssi committed:
Refactoring.

syssi committed:
Missed const updated.

syssi committed:
Incomplete Air Humidifier CA support fixed.

rytilahti commented on homeassistant/components/fan/xiaomi_miio.py:
```diff
@@ -196,50 +437,115 @@ def device_state_attributes(self):
-        """Call an air purifier command handling error messages."""
+    @staticmethod
+    def _extract_value_from_attribute(state, attribute):
+        from enum import Enum
```
Normal imports (not related to backend libraries) should be done on top of the file.

rytilahti commented on homeassistant/components/fan/xiaomi_miio.py:
```diff
@@ -294,70 +602,235 @@ def speed(self):
+            return [mode.name for mode in OperationMode if mode.name != 'Auto']
+
+        # Air Humidifier CA
+        return [mode.name for mode in OperationMode]
```
Is this also valid for non-CA humidifiers? If yes, the comment should be updated.

syssi replied:
I'm using the V1 as new default now. The CA is a very specific device version. My approach was to provide all known operation modes to reduce it later on for specific devices.

rytilahti commented on homeassistant/components/fan/xiaomi_miio.py:
```diff
@@ -294,70 +602,235 @@ def speed(self):
+                self._state_attrs.update(
+                    {key: self._extract_value_from_attribute(state, value)
+                     for key, value in
+                     AVAILABLE_ATTRIBUTES_AIRHUMIDIFIER.items()})
```
This looks like repetition from above (in `__init__`), I think it'd make sense to have a `self._available_attrs` or something initialized in there.

syssi replied:
Good catch!

rytilahti commented on homeassistant/components/fan/xiaomi_miio.py:
```diff
@@ -196,50 +437,115 @@ def device_state_attributes(self):
 
         if result:
             self._state = True
             self._skip_update = True
```
Worth adding a comment wrt `_skip_update` here also?

syssi replied:
The `skip_update` behavior is commented at the `async_update` already. I don't want to introduce a second comment here.

rytilahti reviewed (approved):
Looks great! I added some more comments, other than that I feel this can be merged. Maybe someone else could also do a quick review on this?

Generally speaking it is visible here that the amount of boilerplate required to expose extra services will become a pain to maintain in the future...

syssi committed:
Review incorporated

syssi committed:
The Air Humidifier CA supports the operation mode "auto" - the standard version doesn't

syssi committed:
Attributes are part of the common set already

syssi committed:
Revert "Attributes are part of the common set already"

This reverts commit 40b443eba0e2fc55075479fd540f977fbf4b704a.

syssi committed:
Comment added

syssi committed:
Service description of the set_dry_{on,off} service added

syssi committed:
Typo fixed

rytilahti commented:
As there has been no further requests for changes, let's get this merged for the next release :+1:
